nav2_collision_monitor/src/polygon.cpp

```diff
@@ -300,14 +319,29 @@ bool Polygon::getParameters(std::string & polygon_pub_topic, std::string & footp
       footprint_topic =
         node->get_parameter(polygon_name_ + ".footprint_topic").as_string();
 
-      // This is robot footprint: do not need to get polygon points from ROS parameters.
+      // This is robot footprint: do not need to get polygon points from ROS parameters or topic.
       // It will be set dynamically later.
+      polygon_sub_topic.clear();
       return true;
     } else {
       // Make it empty otherwise
       footprint_topic.clear();
     }
 
+    try {
```

Comment:
There should be an explicit policy that if the polygon points parameter has a polygon in it- we use that and not dynamic from a topic. This seems like it will do both - if the topic exists, it'll set it but also still grab the parameters from the yaml just below. This should be mutually exclusive. So then the polygon class here needs to wait on the first polygon to come in from the topic to be able to meaningfully process data over its region

Reply:
Yep, these parameters already made to be mutually exclusive. The try-catch block handling `polygon_sub_topic` parameter already has a return if `polygon_sub_topic` parameter was set:
```
    try {
      nav2_util::declare_parameter_if_not_declared(".polygon_sub_topic");
      polygon_sub_topic = node->get_parameter(".polygon_sub_topic");
      // Do not need to proceed further, if polygon_sub_topic parameter is defined.
      // Dynamic polygon will be used.
      return true; <- "points" parameter to be ignored in this case.
    } catch (const rclcpp::exceptions::ParameterUninitializedException & ex) {
      // polygon_sub_topic parameter is not defined: initializing polygon with points.
      // Static polygon will be used.
      polygon_sub_topic.clear();
    }

    Handling "points" parameter otherwise -> ...
```


So, we have the following logic:
- If `polygon_sub_topic` was set, polygon subscription will be made and `points` parameter to be ignored.
- If `polygon_sub_topic` was not set, we are reading `points` parameter, which will be mandatory in this case.
- If none (`polygon_sub_topic` and `points`) parameters were not set, getting an error.

The same logic we have for footprint subscriber and its `footprint_topic` parameter for "APPROACH" case, so I doubt, that we should flip the priority here.

Reply:
Got it - the `return` line in L338 I missed, that all makes sense. I think though the opposite policy is most reasonable for default behavior (do you agree?) where if `points` are set, use them. Else, check if there's a topic. If not, error not. My thinking is go from most specific to most general. If something specific is set (points), always use that. 

Also what protections exist to check if polygon is populated from a topic when an iteration is run? Do we have a check if `poly_.size()` is empty and log a warning that it hasn't been set yet by the topic?

Reply:
Okay, I would like to agree and will change the priority of handling for all models in polygons:
- If points set, set polygon shape by them. If them were set incorrectly - fail node. If not set, move to next check
- Subscribe for polygon (or footprint for `APPROACH` model) topic. When receiving, check that message is correct (>= 2 points in there).
- While polygon/footprint was not received - warn a message back to user

> Also what protections exist to check if polygon is populated from a topic when an iteration is run? Do we have a check if poly_.size() is empty and log a warning that it hasn't been set yet by the topic?

In current model not. I will also add it in next commit.

Reply:
Thanks, sounds great!

Reply:
Switched to new logic and added warnings in https://github.com/ros-planning/navigation2/pull/3245/commits/a6939d94c7fbf2f18e0b4be33ea42525e12fd0b6